### PR TITLE
[WebSub] Migrate WebSub module into JVM types/values 

### PR DESCRIPTION
**Pull request by chamil321**
## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Required Balo version update
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [x] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-backend-jvm/src/main/ballerina/compiler_backend_jvm/jvm_method_gen.bal [hunk 1]
```diff
@@ -70,16 +70,16 @@ function genMethodForBallerinaFunction(bir:Function func,
...
-            mv.visitTypeInsn(NEW, typeOwnerClass);
-            mv.visitInsn(DUP);
-            mv.visitMethodInsn(INVOKESPECIAL, typeOwnerClass, "<init>", "()V", false);
-            mv.visitVarInsn(ASTORE, 1);
-            mv.visitLdcInsn(cleanupPackageName(currentPackageName));
-            mv.visitVarInsn(ALOAD, 1);
-            mv.visitMethodInsn(INVOKESTATIC, io:sprintf("%s", VALUE_CREATOR), "addValueCreator",
-                                    io:sprintf("(L%s;L%s;)V", STRING_VALUE, VALUE_CREATOR), false);
-        }
+        string alteredPackageName = "".equalsIgnoreCase(currentPackageName) ? "." : currentPackageName;
```
Comment (MaryamZi) [inline compiler/ballerina-backend-jvm/src/main/ballerina/compiler_backend_jvm/jvm_method_gen.bal]: We don't have to consider case here right 
```suggestion
        string alteredPackageName = currentPackageName == "" ? "." : currentPackageName;
```

File: stdlib/websub/src/main/ballerina/websub/commons.bal [hunk 2]
```diff
@@ -524,19 +512,61 @@ public type WebSubHub object {
...
+        }
+        return registerTopicAtHub(topic);
+    }
 
     # Unregisters a topic in the Ballerina Hub.
     #
     # + topic - The topic to unregister
     # + return - `error` if an error occurred with unregistration
-    public function unregisterTopic(string topic) returns error?;
+    public function unregisterTopic(string topic) returns error?{
```
Comment (MaryamZi) [inline stdlib/websub/src/main/ballerina/websub/commons.bal]:
```suggestion
    public function unregisterTopic(string topic) returns error? {
```

File: stdlib/websub/src/main/ballerina/websub/commons.bal [hunk 3]
```diff
@@ -639,7 +612,7 @@ public type SubscriptionDetails record {|
 |};
 
 function retrieveSubscriberServiceAnnotations(service serviceType) returns SubscriberServiceConfiguration? {
-    any annotData = reflect:getServiceAnnots(serviceType, moduleName = WEBSUB_MODULE_NAME,
+    any annotData = reflect:getServiceAnnotations(serviceType, moduleName = WEBSUB_MODULE_NAME,
                                              ANN_NAME_WEBSUB_SUBSCRIBER_SERVICE_CONFIG);
```
Comment (MaryamZi) [inline stdlib/websub/src/main/ballerina/websub/commons.bal]: We should be able to replace this with 

```ballerina
function retrieveSubscriberServiceAnnotations(service serviceType) returns SubscriberServiceConfiguration? {
    typedesc serviceTypedesc = typeof serviceType;
    return serviceTypedesc.@SubscriberServiceConfig;
} 
```

Can you please check? We might not need the reflect import too.

File: stdlib/websub/src/main/ballerina/websub/h2_hub_persistence.bal [hunk 4]
```diff
@@ -42,12 +42,14 @@ public type H2HubPersistenceStore object {
         self.subscriptionDbClient = subscriptionDbClient;
         var updateResult = self.subscriptionDbClient->update(CREATE_TOPICS_TABLE);
         if (updateResult is error) {
-            panic updateResult;
+            error e = updateResult;
+            panic e;
```
Comment (MaryamZi) [inline stdlib/websub/src/main/ballerina/websub/h2_hub_persistence.bal]: We can merge L45-L46 right?
```suggestion
            panic updateResult;
```

Applicable elsewhere too?

File: stdlib/websub/src/main/ballerina/websub/h2_hub_persistence.bal [hunk 5]
```diff
@@ -65,7 +67,8 @@ public type H2HubPersistenceStore object {
         if (ret is sql:UpdateResult) {
             log:printDebug("Successfully removed " + ret.updatedRowCount + " entries for existing subscription");
         } else {
-            string errCause = <string> ret.detail().message;
+            error e = ret;
```
Comment (MaryamZi) [inline stdlib/websub/src/main/ballerina/websub/h2_hub_persistence.bal]: Again why are we assigning to a new variable?

File: stdlib/websub/src/main/ballerina/websub/hub_client.bal [hunk 6]
```diff
@@ -43,28 +43,78 @@ public type Client client object {
...
+                string payload = result is string ? result : "";
+                error webSubError = error(WEBSUB_ERROR_CODE, message = "Error occurred during topic registration: " + payload);
+                return webSubError;
+            }
+        } else {
+            string errCause = <string> registrationResponse.detail().message;
+            error webSubError = error(WEBSUB_ERROR_CODE, message = "Error sending topic registration request: " + errCause);
+            return webSubError;
+        }
+        return;
```
Comment (MaryamZi) [inline stdlib/websub/src/main/ballerina/websub/hub_client.bal]: We can remove this return. 
```suggestion
```

File: stdlib/websub/src/main/ballerina/websub/subscriber_service_endpoint.bal [hunk 7]
```diff
@@ -41,17 +41,98 @@ public type Listener object {
         return;
     }
 
+    public function __start() returns error? {
+        // TODO: handle data and return error on error
+        self.startWebSubSubscriberServiceEndpoint();
+        self.sendSubscriptionRequests();
+        return;
```
Comment (MaryamZi) [inline stdlib/websub/src/main/ballerina/websub/subscriber_service_endpoint.bal]: Return not needed?

File: stdlib/websub/src/main/java/org/ballerinalang/net/websub/BallerinaWebSubConnectorListener.java [hunk 8]
```diff
@@ -186,30 +183,39 @@ protected void extractPropertiesAndStartResourceExecution(HttpCarbonMessage http
...
-        if (errorStruct != null) {
+        Object returnValue;
+        try {
+            Object[] args = {request, httpResource.getParentService().getBalService()};
+            returnValue = Executor.executeFunction(scheduler, this.getClass().getClassLoader(), WEBSUB, "commons",
+                                                   "processWebSubNotification", args);
+        } catch (BallerinaException e) {
+            log.debug("Signature Validation failed: " + e.getMessage());
+            httpCarbonMessage.setHttpStatusCode(404);
+            throw WebSubUtils.createError(e.getMessage());
```
Comment (MaryamZi) [inline stdlib/websub/src/main/java/org/ballerinalang/net/websub/BallerinaWebSubConnectorListener.java]: When would this fail? If there is an issue with the scheduling/function execution is it?

Also at L220 we seem to be throwing a BallerinaException. In both cases we only need to respond with a 404 right?

File: stdlib/websub/src/main/java/org/ballerinalang/net/websub/broker/BallerinaBroker.java [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Class providing utility methods to interact with the broker operating in in-memory mode.
+ *
+ * @since 0.965.0
+ */
+public class BallerinaBroker {
```
Comment (MaryamZi) [inline stdlib/websub/src/main/java/org/ballerinalang/net/websub/broker/BallerinaBroker.java]: Did we remove the relevant classes from the core? Shall we do so if not?

File: stdlib/websub/src/main/java/org/ballerinalang/net/websub/nativeimpl/RetrieveSubscriptionParameters.java [hunk 10]
```diff
@@ -81,90 +76,70 @@
 
     @Override
     public void execute(Context context) {
-//        Struct subscriberServiceEndpoint = BLangConnectorSPIUtil.getConnectorEndpointStruct(context);
```
Comment (MaryamZi) [inline stdlib/websub/src/main/java/org/ballerinalang/net/websub/nativeimpl/RetrieveSubscriptionParameters.java]: Shall we remove commented out code?

File: stdlib/websub/src/main/ballerina/websub/h2_hub_persistence.bal (hunk 4 above)
Comment (chamil321) [reply stdlib/websub/src/main/ballerina/websub/h2_hub_persistence.bal]: IIRC The type of updateResult inside type guard behaves as a union of errors(error|error) with the recent error type change. In order to panic the error, this casting was needed as the signature of particular function return error?.

File: stdlib/websub/src/main/ballerina/websub/h2_hub_persistence.bal (hunk 5 above)
Comment (chamil321) [reply stdlib/websub/src/main/ballerina/websub/h2_hub_persistence.bal]: In order to get the detail, type should be error. As per my above comment, with the recent error type change, updateResult inside type guard behaves as a union of errors(error|error).

File: stdlib/websub/src/main/java/org/ballerinalang/net/websub/BallerinaWebSubConnectorListener.java (hunk 8 above)
Comment (chamil321) [reply stdlib/websub/src/main/java/org/ballerinalang/net/websub/BallerinaWebSubConnectorListener.java]: Yes, it catches the exceptions which occurs due to ballerina method invocation failure. The message of exception provides a sufficient detail which is being concatenated here.

The thrown exception at L213 will get caught at L127 and leads it to the failure path which is the 404 error response.

File: stdlib/websub/src/main/ballerina/websub/commons.bal (hunk 3 above)
Comment (chamil321) [reply stdlib/websub/src/main/ballerina/websub/commons.bal]: ATM your suggested code results a null pointer at compiler.codegeneretor level. We will change the code once that issue is fixed

File: stdlib/websub/src/main/java/org/ballerinalang/net/websub/nativeimpl/RetrieveSubscriptionParameters.java (hunk 10 above)
Comment (chamil321) [reply stdlib/websub/src/main/java/org/ballerinalang/net/websub/nativeimpl/RetrieveSubscriptionParameters.java]: It's already removed :P

Comment (MaryamZi) [reply stdlib/websub/src/main/java/org/ballerinalang/net/websub/nativeimpl/RetrieveSubscriptionParameters.java]: Sorry, my bad :laughing:

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/values/connector/Executor.java [hunk 11]
```diff
@@ -93,6 +101,78 @@ public static Object executeFunction(Strand strand, ObjectValue service, Attache
...
+     * @param packageName package which the function belongs to
+     * @param className   which the function resides/ or file name
+     * @param methodName  to be invokable unit
+     * @param paramValues to be passed to invokable unit
+     * @return return values
+     */
+    public static Object executeFunction(Scheduler scheduler, ClassLoader classLoader, String packageName,
+                                         String className, String methodName, Object... paramValues) {
+        try {
+            Class<?> clazz = classLoader.loadClass("ballerina." + packageName + "." + className);
```
Comment (wggihan) [inline bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/values/connector/Executor.java]: Shall we change this to support multiple organization without hard coding to ballerina?

File: stdlib/websub/src/main/java/org/ballerinalang/net/websub/BallerinaWebSubConnectorListener.java [hunk 12]
```diff
@@ -220,82 +227,56 @@ private void validateSignature(HttpCarbonMessage httpCarbonMessage, HttpResource
...
      * @return the struct representing the subscriber service endpoint
      */
-    private BMap<String, BValue> getWebSubCaller(HttpResource httpResource, HttpCarbonMessage httpCarbonMessage) {
-        BMap<String, BValue> subscriberServiceServer =
-                createBStruct(httpResource.getBalResource().getResourceInfo().getPackageInfo().getProgramFile(),
-                              WEBSUB_PACKAGE, WEBSUB_SERVICE_LISTENER);
+    private ObjectValue getWebSubCaller(HttpResource httpResource, HttpCarbonMessage httpCarbonMessage,
+                                        MapValue endpointConfig) {
+        ObjectValue httpServiceServer = BallerinaValues.createObjectValue(PROTOCOL_PACKAGE_HTTP, HTTP_LISTENER_ENDPOINT,
+                          9090, endpointConfig); // sending a dummy port here as it gets initialized later - fix
```
Comment (MaryamZi) [inline stdlib/websub/src/main/java/org/ballerinalang/net/websub/BallerinaWebSubConnectorListener.java]: Shall we create an issue for this?

File: stdlib/websub/src/main/java/org/ballerinalang/net/websub/compiler/WebSubServiceCompilerPlugin.java [hunk 13]
```diff
@@ -75,10 +75,10 @@ public void process(ServiceNode serviceNode, List<AnnotationAttachmentNode> anno
             }
         }
 
-        if (webSubAnnotationConfigCount > 1) {
+        if (webSubAnnotationConfigCount == 0) {
             dlog.logDiagnostic(Diagnostic.Kind.ERROR, serviceNode.getPosition(),
-                               "cannot have more than one '" + ANN_NAME_WEBSUB_SUBSCRIBER_SERVICE_CONFIG
-                                       + "' annotation");
+                               "'" + ANN_NAME_WEBSUB_SUBSCRIBER_SERVICE_CONFIG + "' annotation is compulsory");
+            return;
```
Comment (MaryamZi) [inline stdlib/websub/src/main/java/org/ballerinalang/net/websub/compiler/WebSubServiceCompilerPlugin.java]: Shall we add a test for this?

File: stdlib/websub/src/main/java/org/ballerinalang/net/websub/nativeimpl/StopHubService.java [hunk 14]
```diff
@@ -40,20 +42,23 @@
...
         Hub hubInstance = Hub.getInstance();
         if (hubInstance.isStarted()) {
             try {
-                hubInstance.stopHubService(context);
+                hubInstance.stopHubService();
+                return true;
             } catch (BallerinaWebSubException e) {
-                context.setReturnValues(new BBoolean(false));
-                return;
+                log.debug(e.getMessage());
```
Comment (MaryamZi) [inline stdlib/websub/src/main/java/org/ballerinalang/net/websub/nativeimpl/StopHubService.java]: Do we need this?

File: stdlib/websub/src/test/java/org.ballerinalang.net.websub/WebSubHubStartUpTest.java [hunk 15]
```diff
@@ -87,7 +87,8 @@ public void testHubShutdownAndStart() {
 
     @AfterClass
     public void tearDown() {
-        BRunUtil.invoke(result, "stopHub", new BValue[]{hubStartUpObject});
+        //TODO Uncomment following once test suite is migrated to jBal types and values
+//        BRunUtil.invoke(result, "stopHub", new BValue[]{hubStartUpObject});
```
Comment (MaryamZi) [inline stdlib/websub/src/test/java/org.ballerinalang.net.websub/WebSubHubStartUpTest.java]: this needs to change?

File: stdlib/websub/src/test/resources/test-src/hub/test_hub_startup.bal [hunk 16]
```diff
@@ -22,7 +22,9 @@ function startupHub(int hubPort) returns websub:WebSubHub|websub:HubStartedUpErr
     return websub:startHub(new http:Listener(hubPort));
 }
 
-function stopHub(websub:WebSubHub|websub:HubStartedUpError hubStartUpResult) returns boolean {
+//TODO change function to accept websub:WebSubHub|websub:HubStartedUpError hubStartUpResult once test migration is done
```
Comment (MaryamZi) [inline stdlib/websub/src/test/resources/test-src/hub/test_hub_startup.bal]: todo?

File: stdlib/websub/src/main/ballerina/websub/hub_client.bal [hunk 17]
```diff
@@ -74,151 +123,79 @@ public type Client client object {
...
+            }
+        } else {
+            error webSubError = error(WEBSUB_ERROR_CODE, message = "Publish failed for topic [" + topic + "]");
             return webSubError;
         }
-    } else {
-        map<anydata> errorDetail = { message : "Publish failed for topic [" + topic + "]" };
-        error webSubError = error(WEBSUB_ERROR_CODE, errorDetail);
-        return webSubError;
+        return;
```
[the remark this answers is not in the document]
Comment (chamil321) [reply stdlib/websub/src/main/ballerina/websub/hub_client.bal]: We need it as L144 has if condition without an else block

File: stdlib/websub/src/main/java/org/ballerinalang/net/websub/BallerinaWebSubConnectorListener.java (hunk 12 above)
Comment (chamil321) [reply stdlib/websub/src/main/java/org/ballerinalang/net/websub/BallerinaWebSubConnectorListener.java]: +1

File: stdlib/websub/src/test/java/org.ballerinalang.net.websub/WebSubHubStartUpTest.java (hunk 15 above)
Comment (chamil321) [reply stdlib/websub/src/test/java/org.ballerinalang.net.websub/WebSubHubStartUpTest.java]: yes, until we get proper jvm value test suite, we can pass a return value again to another class as it goes through few conversions

File: stdlib/websub/src/test/resources/test-src/hub/test_hub_startup.bal (hunk 16 above)
Comment (chamil321) [reply stdlib/websub/src/test/resources/test-src/hub/test_hub_startup.bal]: Explained by the above comment

File: stdlib/websub/src/main/java/org/ballerinalang/net/websub/compiler/WebSubServiceCompilerPlugin.java (hunk 13 above)
Comment (chamil321) [reply stdlib/websub/src/main/java/org/ballerinalang/net/websub/compiler/WebSubServiceCompilerPlugin.java]: Added

**Pull request merged**